lib/tree.rb

```diff
@@ -2,63 +2,162 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+    @parent = nil
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(logn)
+  # Space Complexity: O(1)
```

Comment:
Since you're using a recursive `add_node` helper method, you are using the system stack and this impacts space complexity.  Since you are making O(log n) recursive calls on a balanced tree this would have a space complexity of O(log n).